### PR TITLE
bugfix: put sandbox into store after created to avoid uncontrolled sandbox container

### PR DESCRIPTION
**Pull request by Starnop**
Signed-off-by: Starnop <starnop@163.com>

### Ⅰ. Describe what this PR did
as title described.

### Ⅱ. Does this pull request fix one issue?

NONE
### Ⅲ. Why don't you add test cases (unit test/integration test)? (你真的觉得不需要加测试吗？)

None.

### Ⅴ. Special notes for reviews

Cherry pick

### CODE REVIEW

File: cri/v1alpha2/cri.go [hunk 1]
```diff
@@ -270,7 +287,10 @@ func (c *CriManager) RunPodSandbox(ctx context.Context, r *runtime.RunPodSandbox
 	defer func() {
 		// If running sandbox failed, clean up the sandbox directory.
 		if retErr != nil {
-			os.RemoveAll(sandboxRootDir)
+			removeErr := os.RemoveAll(sandboxRootDir)
```
Comment (fuweid) [inline cri/v1alpha2/cri.go]: use `if err := xx; err != nil` pattern here?

File: cri/v1alpha2/cri.go [hunk 2]
```diff
@@ -246,12 +246,29 @@ func (c *CriManager) RunPodSandbox(ctx context.Context, r *runtime.RunPodSandbox
...
-		// If running sandbox failed, clean up the container.
 		if retErr != nil {
 			err := c.ContainerMgr.Remove(ctx, id, &apitypes.ContainerRemoveOptions{Volumes: true, Force: true})
 			if err != nil {
-				logrus.Errorf("failed to remove the container when running sandbox failed: %v", err)
+				logrus.Errorf("failed to remove container when running sandbox failed %q: %v", id, err)
+			}
+			// should not remove the sandbox container metadata from sandboxStore
+			// until it was removed by pouchd.
+			removeErr := c.SandboxStore.Remove(id)
```
Comment (fuweid) [inline cri/v1alpha2/cri.go]: use `if err := xx; err != nil pattern` here?

File: cri/v1alpha2/cri.go [hunk 3]
```diff
@@ -306,13 +326,6 @@ func (c *CriManager) RunPodSandbox(ctx context.Context, r *runtime.RunPodSandbox
...
 	}
 
-	sandboxMeta := &SandboxMeta{
-		ID:        id,
-		Config:    config,
-		NetNSPath: netnsPath,
-		Runtime:   config.Annotations[anno.KubernetesRuntime],
-	}
-
 	if _, ok := config.Annotations[anno.LxcfsEnabled]; ok {
```
Comment (fuweid) [inline cri/v1alpha2/cri.go]: could we put it before create sandbox? I think the validation should be checked at first

File: cri/v1alpha2/cri.go (hunk 2 above)
Comment (allencloud) [reply cri/v1alpha2/cri.go]: Yes, it is better to use `if err := xx; err != nil`, and the err would only take effect in the code block. After that, there is no need to add a new naming `removeErr`, `err` is enough.

File: cri/v1alpha2/cri.go [hunk 4]
```diff
@@ -306,21 +333,8 @@ func (c *CriManager) RunPodSandbox(ctx context.Context, r *runtime.RunPodSandbox
...
-	if _, ok := config.Annotations[anno.LxcfsEnabled]; ok {
-		enableLxcfs, err := strconv.ParseBool(config.Annotations[anno.LxcfsEnabled])
-		if err != nil {
-			return nil, err
-		}
-		sandboxMeta.LxcfsEnabled = enableLxcfs
-	}
-
+	// update the metadata of sandbox container after network had been set up successfully.
+	sandboxMeta.NetNSPath = netnsPath
```
Comment (fuweid) [inline cri/v1alpha2/cri.go]: I think this part should be following `c.setupPodNetwork`. If there is no need to setup Pod network, are we need to update the NetNSPath here?

**Comment by Starnop** [comment]
123

**Pull request merged**